### DISCUSSION TITLE
When to write operations in 'Forward' and when to write in 'Message'?

### DISCUSSION THREAD

**Discussion by ShengzheXu** [pyg-team/pytorch_geometric#3706] [Q&A] [answered]
Hey guys. Maybe this is a trivial question.. I'm wondering when should I implement my formulations in the '`forward`' method and when should I implement them in the '`message`' (`propagate`) method?

For example, some tutorials implement the `gcn_norm` in the `forward` method of the `GCNconv` class, while some other tutorials also implement the `gcn_norm` in the `message` method.

Could some one help explain this a bit, like what's the difference of implementation at the two places. Appreciate for any advice.

**Comment by rusty1s** [comment, marked as answer]
There doesn't exist a rule of thumb for this and its mostly user-preference. In general, it is preferred to compute all edge-level representations directly inside `message` since you do not need to take care of gathering features, etc. However, I agree that we are breaking with this pattern by ourselves as well, e.g., in `gcn_norm` :)

**Reply by ShengzheXu** [reply]
Thanks Matthias!
